### ISSUE TITLE
Integrate CompatPowerShellGet into v3

### ISSUE THREAD

**Issue by SydneyhSmith** [PowerShell/PSResourceGet#1064]
### Summary of the new feature / enhancement

Integrate the compatibility layer into v3 so that users can continue to use existing cmdlets

**Comment by ThomasNieto** [comment]
That seems like it will be confusing having the v2 and v3 cmdlets side by side. What is the plan to phase out v2 cmdlet names?

**Comment by ThomasNieto** [comment]
Spoke to @SydneyhSmith during summit. Her plan is to integrate the CompatPowerShellGet in PowerShellGet v3 module and NOT have PowerShellGet v3 have a required dependency on CompatPowerShellGet.

My thoughts on this are:

1. When and how will the old cmdlets be deprecated and removed or will they stay around forever? For example, GA in 7.4 and removed in 7.5.
2. This will cause confusion for new and existing users by having the old and new cmdlets shipped in the same module. You'll get the question when should I use `Install-Module` and when do I use `Install-PSResource`.
3. There should be a way to remove/disable/hide the old cmdlets for users and organizations that have migrated to the new cmdlets.

To summarize if you have to have the old cmdlets built-in PowerShell or available on PowerShell v3 upgrade so it does not break CI then one of two things should be implemented.

1. Have PowerShellGet v3 have a required dependency on CompatPowerShellGet. This is good because it doesn't pollute the `Get-Command -Module PowerShellGet` output with deprecated cmdlets.
2.  If you cannot or will not have the old cmdlets shipped via required dependency then, there should be a way to disable/hide the old cmdlets.

**Comment by SydneyhSmith** [comment]
Adding from PR #1055 
<img width="787" alt="Screenshot 2023-05-03 at 7 11 52 PM" src="https://user-images.githubusercontent.com/43417619/236096247-90dd4079-ba78-483a-b2c5-2716da30be46.png">

**Comment by SydneyhSmith** [comment]
<img width="792" alt="Screenshot 2023-05-03 at 7 12 02 PM" src="https://user-images.githubusercontent.com/43417619/236096266-4263257f-cbc9-46f6-bd63-380934b8cb45.png">

**Comment by ThomasNieto** [comment]
@SydneyhSmith it may be helpful for the team to put the design and goals on the issue before implementation so the community can provide feedback before work has started.

**Comment by SydneyhSmith** [comment]
We have discussed this issue at length internally, and plan to discuss this issue more tomorrow but I wanted to summarize some of the discussion that has occurred so far and provide an update as to options we are considering. 

Not breaking CI/CD (and automation scenarios generally) on day 1 is a goal of ours which left us with three primary options:
1.	Ship compat as a part of PowerShellGet 
2.	Ship compat as a dependency of PowerShellGet 
3.	Rename compat to be PowerShellGet and then take a dependency on v3 which is renamed as PSResourceGet 

Initially we leaned towards option 2 (ship compat as a dependency of PowerShellGet) however this breaks anyone who uses fully qualified names, through our searches on GitHub this includes a lot of automation scenarios. This alone essentially eliminated option 2 for us as not breaking CI/CD was a primary motivator for including compat. Some other factors we have considered: 
•	We are leaning towards supporting v2 syntax for the foreseeable future. We want to have an ability of removing it though should the time come that it is no longer needed. 
•	We view supporting v2 syntax as a compatibility feature, not a place we would continue to make feature investments—future feature investments such as installation of arbitrary nupkgs would only come to Install-PSResource
•	Collecting telemetry on the Gallery side will enable us to know when it may be appropriate to deprecate v2 cmdlets, but this is not something we expect to do over the short term
•	Need to think about what will be preferred approach for going into Windows PowerShell in the future to replace PSGet v1.0.0.1
•	We should make changes to user agent header in v3 for collecting telemetry on Gallery side. Can we do something to find out if it was invoked from compat proxy?
•	There are concerns with marketing of "PSResourceGet", would it be considered a subset or inferior to the "real" PowerShellGet. Would moving people for PSGet to PSResourceGet be difficult?
•	Top priority should be making adoption of PSGetV3 as easy and seamless as possible.
•	PSGetV2/PSGetV3 running side-by-side is not a significant scenario that we should support, PSGetV3 should replace V2 and we should plan to deprecate V2

**Comment by ThomasNieto** [comment]
Thanks for the detailed write-up and helps a lot in regards to the team's thought process.

I did not take into account people who use `fully-qualified` cmdlet names. Is it a lot of users?

> There are concerns with marketing of "PSResourceGet", would it be considered a subset or inferior to the "real" PowerShellGet. Would moving people for PSGet to PSResourceGet be difficult?

I feel this statement is also applies to the new cmdlet names in the same module as the old ones. Which cmdlet is the "real" one and why should I use the new names when the old ones are still there? If you have both legacy and new cmdlets in the same module why even have the renamed cmdlets and just use the legacy ones?

When the RFC was open I was behind using a new module name for this migration since the cmdlet names would be changing.
I believe that if the new user experience is better and has feature parity with v2 then you'll get people to migrate to the new module/cmdlets on their own.

> Top priority should be making adoption of PSGetV3 as easy and seamless as possible.

I also agree with this statement. My main concern is the user experience and having legacy cmdlets and new cmdlets in the same module creates user confusion about when and where to use which.

My opinion is to take option 3 which solves the `fully-qualified` name issue and is backwards compatible with v2 by shipping both `PowerShellGet` module as a compatibility layer while also having a clean break for users with the new module and cmdlets. However, I feel the `PSResourceGet` could have a better name (remove `Get`) 😄

**Comment by SydneyhSmith** [comment]
We had another discussion so I wanted to update with more options. We discussed that because 7.4 is an LTS we want to be very cautious about introducing breaking changes, especially coming in much later than we hoped in the release cycle, so another option we discussed is shipping PowerShellGet v2 and v3 side by side in 7.4 (with no Compat layer). We initially didn't consider this option because it was not possible due to the way that PowerShell is built (and at the time we weren't considering renaming the module), however it is now possible with the option to rename PowerShellGet or not rename PowerShellGet. This would break the fewest people possible, but still allow those who wanted to upgrade to do so. The down side is that this would slow adoption to v3 implementation (although could potentially accelerate adoption of v3 syntax since those who wanted to get the benefits would be more inclined to update scripts, still Compat would be available on the Gallery). Since 7.5 is not LTS and we could ship beginning in Preview 1 we could stop shipping PowerShellGet v2 in 7.5 and introduce one of our compatibility options described above. This would allow us more time to figure out gaps in the proxy functions and help folks with the transition. Personally I am eager to get v3 out there and would love to see it fully replace v2 in 7.4 but want to make the decision that is best for the broader PowerShell community (1 billion monthly sessions).

**Comment by ThomasNieto** [comment]
I like this option the best, just ship both v2 and v3 in PowerShell without the compat layer as its not needed. Users that haven't migrated will still use the exact same version they're use to and for users that have migrated will get the new experience. This also has the added benefit of keeping the module name the same for familiarity.

I would recommend to do the following:

1.  Update PowerShellGet v2 with a warning message when used if the module has been loaded from $PSHOME that the module will no longer be shipped in PowerShell in the future (or X release) and can be installed explicitly if they still need to stay on v2.
2. Since the warning only goes out to people who are using the built-in module they would be impacted while users that have installed it through PowerShellGet even if they upgrade PowerShell, v2 would still stay around in their `AllUsers` or `CurrentUser` path not breaking them.
3. The warning also serves the purpose for CI/CD like how GitHub actions will write a warning if using an out of date image that will be removed in the future.
4. You can also use telemetry to determine when v3 adoption has reached acceptable levels to remove v2 in-box.

I believe with adequate notice with warnings serves is reasonable to notify users to migrate without the need to place the compat layer in PowerShell. You can also use telemetry to determine when v3 adoption has reached acceptable levels to remove v2 in-box. This is still way in the future and we don't know how quickly v3 adoption will take place to warrant shipping the compat layer after v2 has been removed in 7.5/7.6+.

**Comment by alerickson** [comment]
@ThomasNieto, I love the suggestion of adding the warning into v2.  We hadn't thought of adding warnings there in particular, but I think the plan you laid out is a great way to get the deprecation warning out.

**Comment by SydneyhSmith** [comment]
closing this in favor of #1120
